server/hotrod/src/main/java/org/infinispan/server/hotrod/ClientListenerRegistry.java

```diff
@@ -387,43 +402,41 @@ boolean isChannelDisconnected() {
          return !ch.isOpen();
       }
 
-      void sendEvent(byte[] key, byte[] value, long dataVersion, CacheEntryEvent event) {
+      CompletionStage<Void> sendEvent(byte[] key, byte[] value, long dataVersion, CacheEntryEvent event) {
          Events.Event remoteEvent = createRemoteEvent(key, value, dataVersion, event);
          if (isTrace)
             log.tracef("Queue event %s, before queuing event queue size is %d", remoteEvent, eventQueue.size());
 
          EventLoop loop = ch.eventLoop();
-         // We shouldn't be in the event loop, but just in case we can't get stuck blocking on putting into the queue
-         // so we offer and try to catch up on events if possible
-         if (loop.inEventLoop()) {
-            boolean offered = eventQueue.offer(remoteEvent);
-            while (!offered) {
-               // If the event queue is full, we have to try to write some events to free up space since we are in the
-               // event loop and no other thread can drain this queue but us
-               writeEventsIfPossible();
-               // We again try to offer, but if we weren't able to write any events this will not offer - so we
-               // have to wait for the client to catch up to us - we put a little wait to not cause CPU to spin
-               try {
-                  offered = eventQueue.offer(remoteEvent, 1, TimeUnit.MILLISECONDS);
-               } catch (InterruptedException e) {
-                  throw new CacheException(e);
-               }
+         int size = eventSize.incrementAndGet();
+         boolean forceWait = size >= highWaterMark;
+         CompletableFuture<Void> cf;
+         if (forceWait) {
+            if (isTrace) {
+               log.tracef("Pending event size is %s which is forcing %s to delay operation until it is sent", size, event);
             }
-         } else {
-            try {
-               // TODO: replace with a better number
-               if (!eventQueue.offer(remoteEvent, 30, TimeUnit.SECONDS)) {
-                  throw new TimeoutException("Timed out attempting to offer remote event into queue");
-               }
-            } catch (InterruptedException e) {
-               throw new CacheException(e);
+            cf = new CompletableFuture<>();
+            // Have to write to map before queue, as we check in opposite order when sending events
```

Comment:
Which map?

Reply:
Sorry the set, which is a map underneath :), was originally a map and I didn't update the comment.